src/Raven.Client/Http/ClusterRequestExecutor.cs

```diff
@@ -0,0 +1,98 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Raven.Client.Server.Commands;
+using Sparrow.Json;
+
+namespace Raven.Client.Http
+{
+    public class ClusterRequestExecutor : RequestExecutor
+    {
+
+        protected ClusterRequestExecutor(string databaseName, string apiKey) : base(databaseName, apiKey)
+        {
+        }
+
+        public new static ClusterRequestExecutor Create(string[] urls, string databaseName, string apiKey)
+        {
+            var executor = new ClusterRequestExecutor(databaseName, apiKey);
+            executor._firstTopologyUpdate = executor.FirstTopologyUpdate(urls);
+            return executor;
+        }
+
+        public new static RequestExecutor CreateForSingleNode(string url, string databaseName, string apiKey)
+        {
+            return Create(new[] { url }, databaseName, apiKey);
+        }
+
+        protected override async Task FirstTopologyUpdate(string[] initialUrls)
+        {
+            var list = new List<Exception>();
+            foreach (var url in initialUrls)
+            {
+                try
+                {
+                    await GetClusterTopologyAsync(new ServerNode
+                        {
+                            Url = url,
+                            Database = _databaseName
+                        }, Timeout.Infinite)
+                        .ConfigureAwait(false);
+                    return;
+                }
+                catch (Exception e)
+                {
+                    list.Add(e);
+                }
+            }
+
+            //TODO: cache for cluster topology
+
+            _lastKnownUrls = initialUrls;
+
+            throw new AggregateException("Failed to retrieve clsuter topology from all known nodes", list);
```

Comment:
We need to do a match of each node and its error. At least the message of the exception, and then the full details in the `InnerExceptions`